lib/http2/connection.c

```diff
@@ -81,15 +81,40 @@ static void enqueue_goaway(h2o_http2_conn_t *conn, int errnum, h2o_iovec_t addit
     }
 }
 
+static void graceful_shutdown_close_stragglers(h2o_timeout_entry_t *entry)
+{
+    h2o_context_t *ctx = H2O_STRUCT_FROM_MEMBER(h2o_context_t, http2._graceful_shutdown_timeout, entry);
+    h2o_linklist_t *node, *next;
+
+    /* We've sent two GOAWAY frames, close the remaining connections */
+    for (node = ctx->http2._conns.next; node != &ctx->http2._conns; node = next) {
+        h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _conns, node);
+        next = node->next;
+        if (conn->state < H2O_HTTP2_CONN_STATE_IS_CLOSING) {
```

Comment:
If there is a pending write and the client never sends ACKs, then the write would stall for fairly long time without the write completion callback getting called. Eventually the idle timeout will kick in, but until then, I believe that the connection would be kept open.